### PR TITLE
ztest/zdb links to libzoa

### PR DESCRIPTION
**Pull request by manoj-joseph**
Please see the following document for why we are making this change.
https://docs.google.com/document/d/1qp2iPuLm0kwG5nLorJRS9dBJ9rl9gatHutv8W80H_0k/edit#heading=h.2tyuac63vlgs

In brief, zdb and ztest now link to `libzpool` (ZFS kernel code compiled into userland) as well as `libzoa` (the zfs-object-agent library). Since libzoa is written in rust, we have added a C API that can be called from zdb.

I refactored the agent a little. The common code, which used to be called `libzoa` is now called `zoa-common`. `libzoa.so` is the agent library that C programs can link to. The header `libzoa.h` is generated with `cbindgen`. However, this is done manually now. Once I get the package installed on the build VM, we can get this step automated.

The agent initialization code that ztest/zdb need to invoke can be done via the `libzoa` library call `libzoa_init()`. This shares call with the agent's main.rs. The common initialization code (starting the tokio threads as well as logging setup are now in `init.rs`.

Testing:
When compiled without rust, we compile out the support for object store from ztest/zdb.
When rust is present, ztest and zdb communicate with the agent that it links to.

```
delphix@mj-zdb-2:~$ zdb -bccsv -d -Y -e -y  -a https://s3-us-west-2.amazonaws.com -g us-west-2 -B cloudburst-data-2 -f file:///etc/zfs/zpool_credentials 9933524831088661460 | tail
verifying concrete vdev 0, metaslab 0 of 1 ...
     1M:      0      0  6.11M      0      0  14.6M      0      0  6.11M
     2M:      0      0  6.11M      0      0  14.6M      0      0  6.11M
     4M:      0      0  6.11M      0      0  14.6M      0      0  6.11M
     8M:      0      0  6.11M      0      0  14.6M      0      0  6.11M
    16M:      0      0  6.11M      0      0  14.6M      0      0  6.11M

                            capacity   operations   bandwidth  ---- errors ----
description                used avail  read write  read write  read write cksum
9933524831088661460           0  512P    72     0 2.03M     0     0     0     0
  cloudburst-data-2           0  512P    72     0 2.03M     0     0     0     0
delphix@mj-zdb-2:~$ 
```
```
delphix@mj-zdb-2:~$ ztest  -O https://s3-us-west-2.amazonaws.com -A us-west-2 -b cloudburst-data-2 -z file:///etc/zfs/zpool_credentials -VVVVV -T 60
7 datasets, 23 threads, object-store https://s3-us-west-2.amazonaws.com, 60 seconds...

Executing zdb -bccsv -d -Y -e -y  -a https://s3-us-west-2.amazonaws.com -g us-west-2 -B cloudburst-data-2 -f file:///etc/zfs/zpool_credentials 12696753370641087688
verifying concrete vdev 0, metaslab 0 of 1 ...
Verifying deleted livelist entries
Verifying metaslab entries
Dataset mos [META], ID 0, cr_txg 4, 18.5K, 38 objects
<<<<snip>>>>
                            capacity   operations   bandwidth  ---- errors ----
description                used avail  read write  read write  read write cksum
12696753370641087688          0  512P   381     0 4.69M     0     0     0     0
  cloudburst-data-2           0  512P   381     0 4.69M     0     0     0     0
1 killed, 0 completed, 100% kill rate
delphix@mj-zdb-2:~$ 
```

### CODE REVIEW

File: cmd/zdb/zdb.c [hunk 1]
```diff
@@ -767,7 +771,10 @@ usage(void)
 	    "\t%s -E [-A] word0:word1:...:word15\n"
 	    "\t%s -S [-AP] [-e [-V] [-p <path> ...]] [-U <cache>] "
 	    "<poolname>\n\n",
-	    cmdname, cmdname, cmdname, cmdname, cmdname, cmdname, cmdname,
+	    cmdname, cmdname, cmdname, cmdname, cmdname, cmdname,
+#ifdef HAVE_LIBZOA
+	    cmdname,
+#endif
 	    cmdname, cmdname, cmdname, cmdname, cmdname);
```
Comment (ahrens) [inline cmd/zdb/zdb.c:778]: I think using `%1$s` would let us pass `cmdname` exactly once, and use it wherever we need to in the printf string.   If you want to do some cleanup while you're here.

File: cmd/zfs_object_agent/zoa/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "zoa"
+version = "0.1.0"
+authors = ["Manoj Joseph <manoj.joseph@delphix.com>"]
```
Comment (ahrens) [inline cmd/zfs_object_agent/zoa/Cargo.toml:4]: Would you mind using the corporate "authors" as we have in the other Cargo.toml files?
```suggestion
authors = ["Delphix"]
```

File: cmd/zfs_object_agent/zoa/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+pub extern "C" fn libzoa_init(socket_dir_ptr: *const c_char, log_file_ptr: *const c_char) {
+    let socket_dir = unsafe {
+        CStr::from_ptr(socket_dir_ptr)
+            .to_string_lossy()
+            .into_owned()
+    };
+    let log_file = unsafe { CStr::from_ptr(log_file_ptr).to_string_lossy().into_owned() };
+
+    let verbosity: u64 = 2;
+    let file_name: Option<&str> = Some(log_file.as_str());
```
Comment (ahrens) [inline cmd/zfs_object_agent/zoa/src/lib.rs:15]: I don't think you need to specify the types here
```suggestion
    let verbosity = 2;
    let file_name = Some(&log_file);
```

**Review by ahrens — COMMENTED** [review]
Did you consider using cbindgen as a library, rather than as a program?

https://github.com/eqrion/cbindgen/blob/master/docs.md
>There are two ways to use cbindgen: as a standalone program, or as a library (presumably in your build.rs). There isn't really much practical difference, because cbindgen is a simple rust library with no interesting dependencies. Using it as a program means people building your software will need it installed. Using it in your library means people may have to build cbindgen more frequently (e.g. every time they update their rust compiler).

here's another guide: https://michael-f-bryan.github.io/rust-ffi-guide/cbindgen.html

File: cmd/zdb/zdb.c (hunk 1 above)
Comment (manoj-joseph) [reply cmd/zdb/zdb.c]: Fixed

File: cmd/zfs_object_agent/zoa/Cargo.toml (hunk 2 above)
Comment (manoj-joseph) [reply cmd/zfs_object_agent/zoa/Cargo.toml]: Fixed

File: cmd/Makefile.am [hunk 4]
```diff
@@ -18,6 +18,9 @@ if BUILD_RUST
 SUBDIRS += zfs_object_agent
 endif
 
+# ztest and zdb depend on zfs_object_agent.
```
[the remark this answers is not in the document]
Comment (manoj-joseph) [reply cmd/Makefile.am]: ok

**Comment by manoj-joseph** [comment]
> Did you consider using cbindgen as a library, rather than as a program?

I have made the switch to using it as a library.

File: cmd/zfs_object_agent/zoa/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+            .to_string_lossy()
+            .into_owned()
+    };
+    let log_file = unsafe { CStr::from_ptr(log_file_ptr).to_string_lossy().into_owned() };
+
+    let verbosity = 2;
+    let file_name = Some(log_file.as_str());
+
+    init::setup_logging(verbosity, file_name);
+    init::start(socket_dir.as_str(), None);
```
Comment (ahrens) [inline cmd/zfs_object_agent/zoa/src/lib.rs:18]: I think that this works (I'm assuming `socket_dir: String`), due to [deref coersion](https://doc.rust-lang.org/book/ch15-02-deref.html#implicit-deref-coercions-with-functions-and-methods)
```suggestion
    init::start(&socket_dir, None);
```

**Review by ahrens — APPROVED** [review]

**Pull request merged**